// @backstage/plugin-api-docs/dist/esm/GraphQlDefinition-b8afc092.esm.js
import { makeStyles } from '@material-ui/core/styles';
import GraphiQL from 'graphiql';
import 'graphiql/graphiql.css';
import { buildSchema } from 'graphql';
import React from 'react';

const useStyles = makeStyles(() => ({
  root: {
    height: "100%",
    display: "flex",
    flexFlow: "column nowrap"
  },
  graphiQlWrapper: {
    flex: 1,
    "@global": {
      ".graphiql-container": {
        boxSizing: "initial",
        height: "100%",
        minHeight: "600px",
        flex: "1 1 auto"
      }
    }
  }
}));
const GraphQlDefinition = ({ definition }) => {
  const classes = useStyles();
  const schema = buildSchema(definition);
  return /* @__PURE__ */ React.createElement("div", {
    className: classes.root
  }, /* @__PURE__ */ React.createElement("div", {
    className: classes.graphiQlWrapper
  }, /* @__PURE__ */ React.createElement(GraphiQL, {
    fetcher: () => Promise.resolve(null),
    schema,
    docExplorerOpen: true,
    defaultSecondaryEditorOpen: false
  })));
};

export { GraphQlDefinition };
