// @backstage/plugin-search/dist/index.esm.js
export { m as DefaultResultListItem, F as Filters, e as FiltersButton, H as HomePageSearchBar, S as Router, a as SearchBar, b as SearchBarBase, n as SearchBarNext, f as SearchContextProvider, h as SearchFilter, i as SearchFilterNext, d as SearchModal, o as SearchPage, p as SearchPageNext, t as SearchResult, j as SearchResultPager, k as SearchType, l as SidebarSearch, v as SidebarSearchModal, q as plugin, s as searchApiRef, q as searchPlugin, g as useSearch, u as useSearchModal } from './esm/index-6c1324ea.esm.js';
import '@backstage/core-plugin-api';
import '@backstage/errors';
import 'qs';
import 'react';
import '@material-ui/icons/FilterList';
import '@material-ui/core';
import 'react-use/lib/useDebounce';
import '@material-ui/icons/Search';
import '@material-ui/icons/Clear';
import 'react-use/lib/useAsync';
import 'react-use/lib/usePrevious';
import '@material-ui/lab';
import 'react-use/lib/useAsyncFn';
import '@material-ui/icons/Launch';
import '@material-ui/core/styles';
import '@backstage/core-components';
import 'react-text-truncate';
import '@material-ui/icons/ArrowBackIos';
import '@material-ui/icons/ArrowForwardIos';
import 'react-router';
import '@material-ui/core/InputBase';
import '@material-ui/core/IconButton';
import '@backstage/plugin-catalog-react';
import '@backstage/catalog-model';
import 'react-use/lib/useEffectOnce';
import '@material-ui/icons/ExpandMore';
import '@material-ui/icons/FontDownload';
import 'react-router-dom';
