// @backstage/plugin-catalog-common/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const RESOURCE_TYPE_CATALOG_ENTITY = "catalog-entity";
const catalogEntityReadPermission = {
  name: "catalog.entity.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_CATALOG_ENTITY
};
const catalogEntityCreatePermission = {
  name: "catalog.entity.create",
  attributes: {
    action: "create"
  }
};
const catalogEntityDeletePermission = {
  name: "catalog.entity.delete",
  attributes: {
    action: "delete"
  },
  resourceType: RESOURCE_TYPE_CATALOG_ENTITY
};
const catalogEntityRefreshPermission = {
  name: "catalog.entity.refresh",
  attributes: {
    action: "update"
  },
  resourceType: RESOURCE_TYPE_CATALOG_ENTITY
};
const catalogLocationReadPermission = {
  name: "catalog.location.read",
  attributes: {
    action: "read"
  }
};
const catalogLocationCreatePermission = {
  name: "catalog.location.create",
  attributes: {
    action: "create"
  }
};
const catalogLocationDeletePermission = {
  name: "catalog.location.delete",
  attributes: {
    action: "delete"
  }
};

exports.RESOURCE_TYPE_CATALOG_ENTITY = RESOURCE_TYPE_CATALOG_ENTITY;
exports.catalogEntityCreatePermission = catalogEntityCreatePermission;
exports.catalogEntityDeletePermission = catalogEntityDeletePermission;
exports.catalogEntityReadPermission = catalogEntityReadPermission;
exports.catalogEntityRefreshPermission = catalogEntityRefreshPermission;
exports.catalogLocationCreatePermission = catalogLocationCreatePermission;
exports.catalogLocationDeletePermission = catalogLocationDeletePermission;
exports.catalogLocationReadPermission = catalogLocationReadPermission;
