// @backstage/plugin-catalog-common/dist/index.esm.js
const RESOURCE_TYPE_CATALOG_ENTITY = "catalog-entity";
const catalogEntityReadPermission = {
  name: "catalog.entity.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_CATALOG_ENTITY
};
const catalogEntityCreatePermission = {
  name: "catalog.entity.create",
  attributes: {
    action: "create"
  }
};
const catalogEntityDeletePermission = {
  name: "catalog.entity.delete",
  attributes: {
    action: "delete"
  },
  resourceType: RESOURCE_TYPE_CATALOG_ENTITY
};
const catalogEntityRefreshPermission = {
  name: "catalog.entity.refresh",
  attributes: {
    action: "update"
  },
  resourceType: RESOURCE_TYPE_CATALOG_ENTITY
};
const catalogLocationReadPermission = {
  name: "catalog.location.read",
  attributes: {
    action: "read"
  }
};
const catalogLocationCreatePermission = {
  name: "catalog.location.create",
  attributes: {
    action: "create"
  }
};
const catalogLocationDeletePermission = {
  name: "catalog.location.delete",
  attributes: {
    action: "delete"
  }
};

export { RESOURCE_TYPE_CATALOG_ENTITY, catalogEntityCreatePermission, catalogEntityDeletePermission, catalogEntityReadPermission, catalogEntityRefreshPermission, catalogLocationCreatePermission, catalogLocationDeletePermission, catalogLocationReadPermission };
